internal/state/execution.go

```diff
@@ -414,28 +415,51 @@ func buildLastCommitInfo(block *types.Block, store Store, initialHeight int64) a
 	}
 }
 
-// extendedCommitInfo expects an ExtendedCommitInfo, which includes the
-// original precommit votes along with their vote extensions.
-// Then, it populates the corresponding ABCI protobuf structures.
-func extendedCommitInfo(extCommit types.ExtendedCommit) abci.ExtendedCommitInfo {
-	vs := make([]abci.ExtendedVoteInfo, len(extCommit.ExtendedSignatures))
-	for i, ecs := range extCommit.ExtendedSignatures {
+// buildLastExtendedCommitInfo expects an ExtendedCommitInfo, which includes
+// the original precommit votes along with their vote extensions from the last
+// commit. It populates the corresponding ABCI protobuf structures using the
+// validator set obtained from the store at the height of the last extended
+// commit.
+func buildLastExtendedCommitInfo(ec *types.ExtendedCommit, store Store, initialHeight int64) abci.ExtendedCommitInfo {
+	// The current height is ec.Height+1
+	if ec.Height+1 == initialHeight {
+		// There is no extended commit for the initial height.
+		return abci.ExtendedCommitInfo{}
+	}
+
+	valSet, err := store.LoadValidators(ec.Height)
+	if err != nil {
+		panic(fmt.Errorf("failed to load validator set at height %d: %w", ec.Height, err))
+	}
+
```

Comment:
Function `buildLastCommitInfo` checks the lengths of validator set and commit. If they are not equal, it panics. Should we do the same here?